tests/apple-exporter/components/test-class-cover.php

```diff
@@ -6,289 +6,125 @@
-			$this->prophecized_workspace->reveal(),
-			$this->settings,
-			$this->styles,
-			$this->layouts
-		);
-
-		add_filter(
-			'apple_news_cover_json',
-			function ( $json ) {
-				$json['behavior']['type'] = 'background_motion';
-
-				return $json;
-			}
-		);
+		// Create a test post and get JSON for it.
+		$post_id  = self::factory()->post->create();
+		$image_id = $this->get_new_attachment( $post_id, 'Test Caption', 'Test alt text.' );
+		set_post_thumbnail( $post_id, $image_id );
+		$json = $this->get_json_for_post( $post_id );
+		$this->assertEquals( 'header', $json['components'][0]['role'] );
+		$this->assertEquals( 'fancy-layout', $json['components'][0]['layout'] );
 
-		$this->assertEquals(
-			array(
-				'role'       => 'header',
-				'layout'     => 'headerPhotoLayout',
-				'components' => array(
-					array(
-						'role'   => 'photo',
-						'layout' => 'headerPhotoLayout',
-						'URL'    => 'bundle://filename.jpg'
-					)
-				),
-				'behavior'   => array(
-					'type'   => 'background_motion',
-					'factor' => 0.8
-				),
-			),
-			$component->to_array()
-		);
+		// Teardown.
+		remove_filter( 'apple_news_cover_json', [ $this, 'filter_apple_news_cover_json' ] );
 	}
 
 	/**
 	 * Ensures that the lightbox font is set to the same font face as the image caption.
 	 */
-	public function testLightboxFont() {
+	public function test_lightbox_font() {
 		$this->set_theme_settings(
 			[
-				'caption_font'  => 'Menlo-Regular',
-				'cover_caption' => true,
+				'caption_font'         => 'Menlo-Regular',
+				'cover_caption'        => true,
+				'meta_component_order' => [ 'cover' ],
 			]
 		);
 
-		// Create an image and give it a caption.
-		$image_id = $this->get_new_attachment( 0, 'Test Caption!' );
-
-		// Create a test post.
-		$post_id = self::factory()->post->create();
-
-		// Set the featured image for the post.
+		// Create a new post and set the featured image with a caption.
+		$post_id  = self::factory()->post->create();
+		$image_id = $this->get_new_attachment( $post_id, 'Test Caption', 'Test alt text.' );
 		set_post_thumbnail( $post_id, $image_id );
+		$json = $this->get_json_for_post( $post_id );
 
 		// Ensure that the font set on the lightbox is the same as the font set on the caption above.
-		$json = $this->get_json_for_post( $post_id );
 		$this->assertEquals(
 			'Menlo-Regular',
 			$json['components'][0]['components'][0]['caption']['textStyle']['fontName']
 		);
 	}
 
 	/**
-	 * Ensures that the cover caption is not enabled by default, but can be
-	 * enabled via a setting.
+	 * Tests the render method for the component.
 	 */
-	public function testCaptionSetting() {
-		// Create an image and give it a caption.
-		$image_id = $this->get_new_attachment( 0, 'Test Caption!' );
-
-		// Create a test post.
-		$post_id = self::factory()->post->create();
-
-		// Set the featured image for the post.
-		set_post_thumbnail( $post_id, $image_id );
+	public function test_render() {
```

Comment:
I've built some helper functions into Apple News to help with some of this, like the `get_json_for_post` function and the `set_theme_settings` function, and the `get_new_attachment` function. Those are defined in `Apple_News_Testcase`. The `self::factory()->post->create()` function is part of the WP extensions to PHPUnit. Basically, `self::factory()` returns an object that contains a few different factories for creating things, like posts, users, terms, etc. I'm using the post factory to create a new post. The arguments you pass to it are basically the same ones you would pass to `wp_insert_post`. The major difference is that the factory will insert sample content for you into all fields that you don't specify values for. Here, I'm specifying the post content, but not the excerpt, title, etc, so those will all have values, but they will be things like `Title 00001` or whatever.

Function calls like `set_post_thumbnail` and `add_post_meta` are normal WordPress functions that modify the state of the test database and set up data for tests. The WordPress PHPUnit test runner resets the database state between tests, so we can insert test data and query against it, knowing that it will get wiped out between runs.

The `$this->assertEquals` calls are stock PHPUnit and are testing our assumptions about what is being returned.

Basically, the old tests were unit tests that tested the behavior of a specific class. These new tests are end-to-end tests, because we're publishing a new post in the database and getting the JSON for it using the Publish to Apple News plugin, and we're verifying that the JSON it generates is what we expect for the class that is responsible for the piece of the JSON that we're testing.

I'd be happy to do a deeper dive into this with you, if you'd like.